README.md
# sliver_bar_chart

<a href="https://flutter.dev/"><img src="https://img.shields.io/badge/flutter-website-deepskyblue.svg" alt="Flutter Website"></a>
<a href="https://dart.dev"><img src="https://img.shields.io/badge/dart-website-deepskyblue.svg" alt="Dart Website"></a>
<a href="https://developer.android.com" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-Android-deepskyblue">
</a>
<a href="https://developer.apple.com/ios/" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-iOS-deepskyblue">
</a>
<a href="" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-Web-deepskyblue">
</a>
<a href="" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-Mac-deepskyblue">
</a>
<a href="" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-Linux-deepskyblue">
</a>
<a href="" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/badge/platform-Windows-deepskyblue">
</a>
<a href=""><img src="https://app.codacy.com/project/badge/Grade/dc683c9cc61b499fa7cdbf54e4d9ff35"/></a>
<a href="https://github.com/Mindinventory/sliver_bar_chart/blob/master/LICENSE" style="pointer-events: stroke;" target="_blank">
<img src="https://img.shields.io/github/license/Mindinventory/sliver_bar_chart"></a>
<a href="https://pub.dev/packages/sliver_bar_chart"><img src="https://img.shields.io/pub/v/sliver_bar_chart?color=as&label=sliver_bar_chart&logo=as1&logoColor=blue&style=social"></a>
<a href="https://github.com/Mindinventory/sliver_bar_chart"><img src="https://img.shields.io/github/stars/Mindinventory/sliver_bar_chart?style=social" alt="MIT License"></a>

A package that supports Bar Chart in a Flutter Sliver. This Package allow us to add Bar Chart in
Sliver and its set a Bar Chart as a Header on Slivers Scroll.

## Key Features

* easy way to add Bar Chart in a Sliver.
* used to set a Bar Chart as a Sliver Header.

# Preview

![sliver_bar_chart](https://github.com/Mindinventory/sliver_bar_chart/blob/master/assets/sliver_bar_chart.gif)

## Basic Usage

Import it to your project file

```
import 'package:sliver_bar_chart/sliver_bar_chart.dart';
```

And add it in its most basic form like it:

```
SliverBarChart(
  barWidget: BarChartWidget(
    minHeight: 100.0,
    maxHeight: 1000.0,
    barValues: [
      BarChartData(
        x: '2022',
        y: 500.0,
        barColor: Colors.blue,
      ),
    ],
    isScrolling: true,
  ),
);
```

### Required parameters of SliverBarChart
------------

| Parameter |  Description  |
| ------------ |  ------------ |
| BarChartWidget barWidget | Used to paint bar chart using custom painter |

### Optional parameters of SliverBarChart
------------

| Parameter |  Default | Description  |
| ------------ | ------------ | ------------ |
| bool restrain | false | Whether to stick the header to the start of the viewport once it has reached its minimum size |
| bool scrolling | - | Whether sliver is scrolling or not |
| List<double> heightArray | [50.0, 78.0, 90.0, 67.0, 36.0] | The array of height for generating the bars of the Bar Chart |
| double maxWidth | 75.0 | The maximum width dimensions of the sliver |
| double minHeight | 30.0 | The minimum height dimensions of the sliver |
| double maxHeight | 175.0 | The maximum height dimensions of the sliver |
| bool fluctuating | false | Whether the header should immediately grow again if the user reverses scroll direction |

### Required parameters of BarChartWidget
------------

| Parameter |  Description  |
| ------------ |  ------------ |
| double minHeight | The minimum height dimensions of the BarChartWidget |
| double maxHeight | The maximum height dimensions of the BarChartWidget |
| List<BarChartData> barValues | hold the list of bar chart data |
| bool isScrolling | Whether sliver is scrolling or not |

### Optional parameters of BarChartWidget
------------

| Parameter |  Default | Description  |
| ------------ | ------------ | ------------ |
| int yAxisIntervalCount | 5 | Used to set interval point on Y axis |
| double xAxisTextRotationAngle | 0.0 | Used to set X axis text on a rotation angle in case of larger text |

### Required parameters of BarChartData Model Class
------------

| Parameter |  Description  |
| ------------ |  ------------ |
| String x | a text that visible on Bar Chart X axis |
| double y | a value that visible on Bar Chart Y axis |

### Optional parameters of BarChartData Model Class
------------

| Parameter |  Default | Description  |
| ------------ | ------------ | ------------ |
| Color barColor | Colors.grey | used to set color of the bar |

### Guideline for contributors
------------

* Contribution towards our repository is always welcome, we request contributors to create a pull
  request for development.

### Guideline to report an issue/feature request
------------
It would be great for us if the reporter can share the below things to understand the root cause of
the issue.

* Library version
* Code snippet
* Logs if applicable
* Device specification like (Manufacturer, OS version, etc)
* Screenshot/video with steps to reproduce the issue
* Library used

LICENSE!
------------
**sliver_bar_chart**
is [MIT-licensed.](https://github.com/Mindinventory/sliver_bar_chart/blob/master/LICENSE)

Let us know!
------------
We’d be really happy if you send us links to your projects where you use our component. Just send an
email to sales@mindinventory.com And do let us know if you have any questions or suggestion
regarding our work.

CHANGELOG.md
[0.0.1] - 2022-01-04

* Initial Release.

example/lib/main.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:sliver_bar_chart/sliver_bar_chart.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: MyHomePage(),
      debugShowCheckedModeBanner: false,
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {
  final ScrollController _scrollController = ScrollController();
  late final List<BarChartData> _barValues;

  final double _minHeight = AppBar().preferredSize.height;
  final double _xAxisTextRotationAngle = 180.0;
  final int _yAxisIntervalCount = 8;
  double _maxHeight = 500.0;
  double _maxWidth = 10.0;

  final bool _restrain = true;
  final bool _fluctuating = false;
  bool _isScrolling = true;

  final int _sliverListChildCount = 20;

  @override
  void initState() {
    super.initState();
    _setupBarChartValues();

    _scrollController.addListener(() {
      setState(() {
        if (_scrollController.offset.roundToDouble() < 100.0) {
          _maxHeight = 500.0;
          _maxWidth = 10.0;
          _isScrolling = true;
        } else {
          if (_scrollController.offset.roundToDouble() >= 400.0) {
            _maxWidth = _scrollController.offset - 10;
          }
          _isScrolling = false;
        }
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: CustomScrollView(
          controller: _scrollController,
          slivers: [
            SliverBarChart(
              restrain: _restrain,
              fluctuating: _fluctuating,
              minHeight: _minHeight,
              maxHeight: _maxHeight,
              maxWidth: _maxWidth,
              barWidget: BarChartWidget(
                maxHeight: _maxHeight,
                minHeight: _minHeight,
                barValues: _barValues,
                isScrolling: _isScrolling,
                yAxisIntervalCount: _yAxisIntervalCount,
                xAxisTextRotationAngle: _xAxisTextRotationAngle,
              ),
            ),
            SliverList(
              delegate: SliverChildBuilderDelegate(
                (BuildContext _, int index) => Container(
                  color: index.isOdd ? Colors.white : Colors.black12,
                  child: Center(
                    child: Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: Text(
                        '$index',
                        textScaleFactor: 3.0,
                      ),
                    ),
                  ),
                ),
                childCount: _sliverListChildCount,
              ),
            ),
          ],
        ),
      ),
      resizeToAvoidBottomInset: false,
    );
  }

  void _setupBarChartValues() {
    _barValues = [
      BarChartData(
        x: '2014',
        y: 500.0,
        barColor: _getRandomColor(),
      ),
      BarChartData(
        x: '2015',
        y: 800.0,
        barColor: _getRandomColor(),
      ),
      BarChartData(
        x: '2016',
        y: 600.0,
        barColor: _getRandomColor(),
      ),
      BarChartData(
        x: '2017',
        y: 900.0,
        barColor: _getRandomColor(),
      ),
      BarChartData(
        x: '2018',
        y: 1000.0,
        barColor: _getRandomColor(),
      ),
      BarChartData(
        x: '2019',
        y: 700.0,
        barColor: _getRandomColor(),
      ),
      BarChartData(
        x: '2020',
        y: 500.0,
        barColor: _getRandomColor(),
      ),
      BarChartData(
        x: '2021',
        y: 300.0,
        barColor: _getRandomColor(),
      ),
    ];
  }

  Color _getRandomColor() {
    return Color(
      (math.Random().nextDouble() * 0xFFFFFF).toInt(),
    ).withOpacity(1.0);
  }
}


lib/src/widgets/bar_widget.dart
import 'package:flutter/material.dart';

import '../models/chart_data.dart';
import '../painters/bar_chart_painter.dart';

/// Bar Widget that paint bar chart using custom painter
class BarChartWidget extends StatefulWidget {
  const BarChartWidget({
    required this.minHeight,
    required this.maxHeight,
    required this.barValues,
    required this.isScrolling,
    this.yAxisIntervalCount = 5,
    this.xAxisTextRotationAngle = 0.0,
    Key? key,
  }) : super(key: key);

  final double minHeight;
  final double maxHeight;
  final List<BarChartData> barValues;
  final bool isScrolling;
  final int yAxisIntervalCount;
  final double xAxisTextRotationAngle;

  @override
  _BarChartWidgetState createState() => _BarChartWidgetState();
}

class _BarChartWidgetState extends State<BarChartWidget> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      resizeToAvoidBottomInset: false,
      body: SafeArea(
        child: AnimatedContainer(
          padding: EdgeInsets.all(widget.isScrolling ? 40.0 : 0.0),
          duration: const Duration(milliseconds: 500),
          child: CustomPaint(
            painter: BarChartPainter(
              minHeight: widget.minHeight,
              barValues: widget.barValues,
              isScrolling: widget.isScrolling,
              yAxisIntervalCount: (widget.yAxisIntervalCount <= 0)
                  ? 1
                  : widget.yAxisIntervalCount,
              xAxisTextRotationAngle: widget.xAxisTextRotationAngle,
            ),
            child: SizedBox(
              height: MediaQuery.of(context).size.height,
              width: MediaQuery.of(context).size.width,
            ),
          ),
        ),
      ),
    );
  }
}


lib/src/painters/bar_chart_painter.dart
import 'dart:ui';

import 'package:flutter/material.dart';

import '../models/chart_data.dart';

/// Custom Painter Used to Draw Bar Chart
class BarChartPainter extends CustomPainter {
  /// Values
  final List<BarChartData> barValues;
  final int yAxisIntervalCount;
  final bool isScrolling;
  late final double _maxValue;
  double _sumOfValues = 0.0;

  /// Height
  late final double _doubleMinHeight;
  final double minHeight;

  /// Width
  double _lastWidth = 0;
  late final double _barWidth;
  late final double _halfBarWidth;
  late final double _availableWidth;
  final List<double> _projectedBarWidth = [];
  final List<double> _yAxisIntervalsText = [];
  late final double xAxisTextRotationAngle;

  /// Margin
  final double _margin = 5;
  late final double _totalMargin;

  /// Bar Painters
  final Paint _barPaint = Paint()..style = PaintingStyle.stroke;
  final Paint _barAxisPaint = Paint()
    ..style = PaintingStyle.stroke
    ..strokeWidth = 1.0
    ..color = Colors.black;

  /// Text Painters
  final TextPainter _xAxisTextPainter = TextPainter(
    textDirection: TextDirection.ltr,
  );

  final TextPainter _yAxisTextPainter = TextPainter(
    textDirection: TextDirection.rtl,
  );

  BarChartPainter({
    required this.minHeight,
    required this.barValues,
    required this.isScrolling,
    required this.yAxisIntervalCount,
    required this.xAxisTextRotationAngle,
  }) {
    /// Calculate total margin we need to add around and between bars
    _totalMargin = (_margin * (barValues.length + 1));

    /// Double of minimum height, which we use to start animation.
    _doubleMinHeight = minHeight * 2;

    /// Find max bar value
    _maxValue = barValues
        .reduce((currentData, nextData) =>
            currentData.y > nextData.y ? currentData : nextData)
        .y;

    /// Find total of values
    for (var barValue in barValues) {
      _sumOfValues += barValue.y;
    }

    final List<double> tempTexts = [];

    /// Setting Up Y axis text intervals
    for (var i = 0; i < yAxisIntervalCount; i++) {
      tempTexts.add(((_maxValue / yAxisIntervalCount) * i));
    }
    tempTexts.add(_maxValue);
    _yAxisIntervalsText.clear();
    _yAxisIntervalsText.addAll(tempTexts.reversed.toList());
  }

  @override
  void paint(Canvas canvas, Size size) {
    /// if width of the view is changed then and then only recalculate this.
    if (size.width != _lastWidth) {
      /// find standard bar width
      _barWidth = (size.width - _totalMargin) / barValues.length;

      /// find half bar height
      _halfBarWidth = _barWidth / 2;

      /// calculate projected bar width
      _availableWidth = size.width - _totalMargin;

      /// calculate projected bar width when our view is fully collapsed
      for (var barValue in barValues) {
        _projectedBarWidth.add((barValue.y * _availableWidth) / _sumOfValues);
      }

      _lastWidth = size.width;
    }

    final List<double> _animatedBarWidths = [];

    for (var barIndex = 0; barIndex < barValues.length; barIndex++) {
      _barPaint.color = barValues[barIndex].barColor;

      /// Check here height if height less then double of minimum height,
      /// if so do some different rendering
      if (size.height <= _doubleMinHeight) {
        /// calc fraction between min height and double of minimum height.(0...0.5...1)
        double fraction = 1 - ((size.height - minHeight) / minHeight);

        /// BAR HEIGHT CALC
        /// take current bar height and do transition to full height
        double barHeight = lerpDouble(
            _doubleMinHeight * barValues[barIndex].y / _maxValue,
            size.height,
            fraction)!;

        /// BAR WIDTH CALC
        double particularBarWidth =
            lerpDouble(_barWidth, _projectedBarWidth[barIndex], fraction)!;
        _barPaint.strokeWidth = particularBarWidth;

        /// BAR POSITION CALC
        double barPosition = (_margin * (barIndex + 1)) +
            (_animatedBarWidths.isNotEmpty
                ? _animatedBarWidths.reduce((currentBarWidth, nextBarWidth) =>
                    currentBarWidth + nextBarWidth)
                : 0) +
            (particularBarWidth / 2);

        _animatedBarWidths.add(particularBarWidth);

        canvas.drawLine(
          Offset(barPosition, size.height),
          Offset(barPosition, size.height - barHeight),
          _barPaint,
        );
      } else {
        _barPaint.strokeWidth = _barWidth;
        double barPosition = _margin +
            ((_barWidth * (barIndex + 1)) - _halfBarWidth) +
            (_margin * barIndex);

        /// calculate bar height according to available height of view
        double barHeight =
            (size.height * barValues[barIndex].y / _maxValue).roundToDouble();

        /// Drawing Bar
        _barPaint.strokeWidth = _barWidth;

        canvas.drawLine(
          Offset(barPosition, size.height),
          Offset(barPosition, size.height - barHeight),
          _barPaint,
        );

        if (isScrolling) {
          /// Drawing X axis
          canvas.drawLine(
            Offset(-0.5, size.height),
            Offset(size.width, size.height),
            _barAxisPaint,
          );

          /// Drawing X axis text
          _xAxisTextPainter.text = TextSpan(
            text: barValues[barIndex].x,
            style: const TextStyle(
              color: Colors.black,
              fontSize: 10.0,
              height: 2.0,
            ),
          );

          _xAxisTextPainter.layout(
            maxWidth: size.width,
          );

          final offsetX = Offset((barPosition - (_margin * 2)), size.height);
          if (xAxisTextRotationAngle != 0.0) {
            canvas.save();
            canvas.translate(offsetX.dx, offsetX.dy);
            canvas.rotate(xAxisTextRotationAngle / 360.0);
            _xAxisTextPainter.paint(canvas, Offset(_margin * 2, 0.0));
            canvas.restore();
          } else {
            _xAxisTextPainter.paint(canvas, offsetX);
          }

          /// Drawing Y axis
          canvas.drawLine(
            Offset(0, size.height),
            const Offset(0, -0.5),
            _barAxisPaint,
          );

          /// Drawing Y axis text based on interval count
          for (var textIndex = 0;
              textIndex <= yAxisIntervalCount;
              textIndex++) {
            double yAxisInterval =
                (size.height / yAxisIntervalCount * textIndex);

            _yAxisTextPainter.text = TextSpan(
              text: _yAxisIntervalsText[textIndex].toInt().toString(),
              style: const TextStyle(
                color: Colors.black,
                fontSize: 10.0,
                height: 2.0,
              ),
            );

            _yAxisTextPainter.layout(
              maxWidth: size.width,
            );

            final offsetY = Offset(-30, yAxisInterval);
            _yAxisTextPainter.paint(canvas, offsetY);
          }
        }
      }
    }
  }

  /// Calculating percentage for width calculate
  getBarWidthPercentage(totalSum, double perticularGraphvalue) {
    return (100 * perticularGraphvalue) / totalSum;
  }

  /// Managing Height values 160 to 80
  manageHeight(double perticularGraphvalue, double height) {
    return (height / perticularGraphvalue) + (height - (height / 2));
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
}


lib/src/widgets/custom_sliver_component.dart
import 'package:flutter/material.dart';

/// Delegate for the sliver component.
class CustomSliverComponentDelegate extends SliverPersistentHeaderDelegate {
  final Widget childWidget;
  final double? minHeight;
  final bool? scrolling;
  final double? maxHeight;
  final double? maxWidth;
  final List heightArray;

  CustomSliverComponentDelegate({
    required this.heightArray,
    required this.childWidget,
    this.minHeight,
    this.scrolling,
    this.maxHeight,
    this.maxWidth,
  });

  @override
  Widget build(
    BuildContext context,
    double shrinkOffset,
    bool overlapsContent,
  ) {
    return childWidget;
  }

  @override
  double get maxExtent => maxHeight ?? 0.0;

  @override
  double get minExtent => minHeight ?? 0.0;

  @override
  bool shouldRebuild(SliverPersistentHeaderDelegate oldDelegate) {
    return true;
  }
}


lib/src/widgets/sliver_bar_chart.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

import 'bar_widget.dart';
import 'custom_sliver_component.dart';

/// A sliver whose size varies when the sliver is scrolled to the edge
/// of the viewport opposite the sliver's [GrowthDirection].
///
/// In the normal case of a [CustomScrollView] with no centered sliver, this
/// sliver will vary its size when scrolled to the leading edge of the viewport.
///
/// This is the layout primitive that [SliverAppBar] uses for its
/// shrinking/growing effect.
class SliverBarChart extends StatefulWidget {
  /// Creates a sliver that varies its size when it is scrolled to the start of
  /// a viewport.
  ///
  /// The [delegate], [restrain], and [fluctuating] arguments must not be null.
  const SliverBarChart({
    Key? key,
    this.restrain = false,
    this.scrolling,
    this.heightArray = const [50.0, 78.0, 90.0, 67.0, 36.0],
    this.maxWidth = 75.0,
    this.minHeight = 30.0,
    this.maxHeight = 175.0,
    this.fluctuating = false,
    required this.barWidget,
  }) : super(key: key);

  /// Configuration for the sliver's layout.
  ///
  /// The delegate provides the following information:
  ///
  ///  * The minimum and maximum dimensions of the sliver.
  ///
  ///  * The builder for generating the widgets of the sliver.
  ///
  ///  * The instructions for snapping the scroll offset, if [fluctuating] is true.

  final double maxWidth;
  final bool? scrolling;
  final double maxHeight;
  final double minHeight;
  final List<double> heightArray;
  final BarChartWidget barWidget;

  /// Whether to stick the header to the start of the viewport once it has
  /// reached its minimum size.
  ///
  /// If this is false, the header will continue scrolling off the screen after
  /// it has shrunk to its minimum extent.
  final bool restrain;

  /// Whether the header should immediately grow again if the user reverses
  /// scroll direction.
  ///
  /// If this is false, the header only grows again once the user reaches the
  /// part of the viewport that contains the sliver.
  ///
  /// The [delegate]'s [CustomSliverComponentDelegate.snapConfiguration] is
  /// ignored unless [fluctuating] is true.
  final bool fluctuating;

  @override
  _SliverBarChartState createState() => _SliverBarChartState();
}

class _SliverBarChartState extends State<SliverBarChart> {
  late CustomSliverComponentDelegate delegate;

  @override
  Widget build(BuildContext context) {
    delegate = CustomSliverComponentDelegate(
      childWidget: widget.barWidget,
      minHeight: widget.minHeight,
      scrolling: widget.scrolling,
      maxHeight: widget.maxHeight,
      maxWidth: widget.maxWidth,
      heightArray: widget.heightArray,
    );

    if (widget.fluctuating && widget.restrain) {
      return _SliverFloatingRestrainPersistentHeader(delegate: delegate);
    }
    if (widget.restrain) {
      return _SliverRestrainPersistentHeader(delegate: delegate);
    }
    if (widget.fluctuating) {
      return _SliverFluctuatingPersistentHeader(delegate: delegate);
    }
    return _SliverScrollingPersistentHeader(delegate: delegate);
  }

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(
      DiagnosticsProperty<CustomSliverComponentDelegate>(
        'delegate',
        delegate,
      ),
    );
    final List<String> flags = <String>[
      if (widget.restrain) 'restrain',
      if (widget.fluctuating) 'floating',
    ];
    if (flags.isEmpty) flags.add('normal');
    properties.add(IterableProperty<String>('mode', flags));
  }
}

class _SliverPersistentHeaderElement extends RenderObjectElement {
  _SliverPersistentHeaderElement(
    _SliverPersistentHeaderRenderObjectWidget widget,
  ) : super(widget);

  @override
  _SliverPersistentHeaderRenderObjectWidget get widget =>
      super.widget as _SliverPersistentHeaderRenderObjectWidget;

  @override
  _RenderSliverPersistentHeaderForWidgetsMixin get renderObject =>
      super.renderObject as _RenderSliverPersistentHeaderForWidgetsMixin;

  @override
  void mount(Element? parent, Object? newSlot) {
    super.mount(parent, newSlot);
    renderObject._element = this;
  }

  @override
  void unmount() {
    super.unmount();
    renderObject._element = null;
  }

  @override
  void update(_SliverPersistentHeaderRenderObjectWidget newWidget) {
    final _SliverPersistentHeaderRenderObjectWidget oldWidget = widget;
    super.update(newWidget);
    final CustomSliverComponentDelegate newDelegate = newWidget.delegate;
    final CustomSliverComponentDelegate oldDelegate = oldWidget.delegate;
    if (newDelegate != oldDelegate &&
        (newDelegate.runtimeType != oldDelegate.runtimeType ||
            newDelegate.shouldRebuild(oldDelegate))) {
      renderObject.triggerRebuild();
    }
  }

  @override
  void performRebuild() {
    super.performRebuild();
    renderObject.triggerRebuild();
  }

  Element? child;

  void _build(double shrinkOffset, bool overlapsContent) {
    owner!.buildScope(this, () {
      child = updateChild(
        child,
        widget.delegate.build(
          this,
          shrinkOffset,
          overlapsContent,
        ),
        null,
      );
    });
  }

  @override
  void forgetChild(Element child) {
    assert(child == this.child);
    this.child = null;
    super.forgetChild(child);
  }

  @override
  void insertRenderObjectChild(covariant RenderBox child, Object? slot) {
    assert(renderObject.debugValidateChild(child));
    renderObject.child = child;
  }

  @override
  void moveRenderObjectChild(
      covariant RenderObject child, Object? oldSlot, Object? newSlot) {
    assert(false);
  }

  @override
  void removeRenderObjectChild(covariant RenderObject child, Object? slot) {
    renderObject.child = null;
  }

  @override
  void visitChildren(ElementVisitor visitor) {
    if (child != null) visitor(child!);
  }
}

abstract class _SliverPersistentHeaderRenderObjectWidget
    extends RenderObjectWidget {
  const _SliverPersistentHeaderRenderObjectWidget({
    Key? key,
    required this.delegate,
  }) : super(key: key);

  final CustomSliverComponentDelegate delegate;

  @override
  _SliverPersistentHeaderElement createElement() =>
      _SliverPersistentHeaderElement(this);

  @override
  _RenderSliverPersistentHeaderForWidgetsMixin createRenderObject(
      BuildContext context);

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder description) {
    super.debugFillProperties(description);
    description.add(
      DiagnosticsProperty<CustomSliverComponentDelegate>(
        'delegate',
        delegate,
      ),
    );
  }
}

mixin _RenderSliverPersistentHeaderForWidgetsMixin
    on RenderSliverPersistentHeader {
  _SliverPersistentHeaderElement? _element;

  @override
  double get minExtent => _element!.widget.delegate.minExtent;

  @override
  double get maxExtent => _element!.widget.delegate.maxExtent;

  @override
  void updateChild(double shrinkOffset, bool overlapsContent) {
    assert(_element != null);
    _element!._build(shrinkOffset, overlapsContent);
  }

  @protected
  void triggerRebuild() {
    markNeedsLayout();
  }
}

class _SliverScrollingPersistentHeader
    extends _SliverPersistentHeaderRenderObjectWidget {
  const _SliverScrollingPersistentHeader({
    Key? key,
    required CustomSliverComponentDelegate delegate,
  }) : super(
          key: key,
          delegate: delegate,
        );

  @override
  _RenderSliverPersistentHeaderForWidgetsMixin createRenderObject(
      BuildContext context) {
    return _RenderSliverScrollingPersistentHeaderForWidgets(
      stretchConfiguration: delegate.stretchConfiguration,
    );
  }
}

class _RenderSliverScrollingPersistentHeaderForWidgets
    extends RenderSliverScrollingPersistentHeader
    with _RenderSliverPersistentHeaderForWidgetsMixin {
  _RenderSliverScrollingPersistentHeaderForWidgets({
    RenderBox? child,
    OverScrollHeaderStretchConfiguration? stretchConfiguration,
  }) : super(
          child: child,
          stretchConfiguration: stretchConfiguration,
        );
}

class _SliverRestrainPersistentHeader
    extends _SliverPersistentHeaderRenderObjectWidget {
  const _SliverRestrainPersistentHeader({
    Key? key,
    required CustomSliverComponentDelegate delegate,
  }) : super(
          key: key,
          delegate: delegate,
        );

  @override
  _RenderSliverPersistentHeaderForWidgetsMixin createRenderObject(
      BuildContext context) {
    return _RenderSliverPinnedPersistentHeaderForWidgets(
      stretchConfiguration: delegate.stretchConfiguration,
      showOnScreenConfiguration: delegate.showOnScreenConfiguration,
    );
  }
}

class _RenderSliverPinnedPersistentHeaderForWidgets
    extends RenderSliverPinnedPersistentHeader
    with _RenderSliverPersistentHeaderForWidgetsMixin {
  _RenderSliverPinnedPersistentHeaderForWidgets({
    RenderBox? child,
    OverScrollHeaderStretchConfiguration? stretchConfiguration,
    PersistentHeaderShowOnScreenConfiguration? showOnScreenConfiguration,
  }) : super(
          child: child,
          stretchConfiguration: stretchConfiguration,
          showOnScreenConfiguration: showOnScreenConfiguration,
        );
}

class _SliverFluctuatingPersistentHeader
    extends _SliverPersistentHeaderRenderObjectWidget {
  const _SliverFluctuatingPersistentHeader({
    Key? key,
    required CustomSliverComponentDelegate delegate,
  }) : super(
          key: key,
          delegate: delegate,
        );

  @override
  _RenderSliverPersistentHeaderForWidgetsMixin createRenderObject(
      BuildContext context) {
    return _RenderSliverFloatingPersistentHeaderForWidgets(
      vsync: delegate.vsync,
      snapConfiguration: delegate.snapConfiguration,
      stretchConfiguration: delegate.stretchConfiguration,
      showOnScreenConfiguration: delegate.showOnScreenConfiguration,
    );
  }

  @override
  void updateRenderObject(BuildContext context,
      _RenderSliverFloatingPersistentHeaderForWidgets renderObject) {
    renderObject.vsync = delegate.vsync;
    renderObject.snapConfiguration = delegate.snapConfiguration;
    renderObject.stretchConfiguration = delegate.stretchConfiguration;
    renderObject.showOnScreenConfiguration = delegate.showOnScreenConfiguration;
  }
}

class _RenderSliverFloatingPinnedPersistentHeaderForWidgets
    extends RenderSliverFloatingPinnedPersistentHeader
    with _RenderSliverPersistentHeaderForWidgetsMixin {
  _RenderSliverFloatingPinnedPersistentHeaderForWidgets({
    RenderBox? child,
    required TickerProvider? vsync,
    FloatingHeaderSnapConfiguration? snapConfiguration,
    OverScrollHeaderStretchConfiguration? stretchConfiguration,
    PersistentHeaderShowOnScreenConfiguration? showOnScreenConfiguration,
  }) : super(
          child: child,
          vsync: vsync,
          snapConfiguration: snapConfiguration,
          stretchConfiguration: stretchConfiguration,
          showOnScreenConfiguration: showOnScreenConfiguration,
        );
}

class _SliverFloatingRestrainPersistentHeader
    extends _SliverPersistentHeaderRenderObjectWidget {
  const _SliverFloatingRestrainPersistentHeader({
    Key? key,
    required CustomSliverComponentDelegate delegate,
  }) : super(
          key: key,
          delegate: delegate,
        );

  @override
  _RenderSliverPersistentHeaderForWidgetsMixin createRenderObject(
      BuildContext context) {
    return _RenderSliverFloatingPinnedPersistentHeaderForWidgets(
      vsync: delegate.vsync,
      snapConfiguration: delegate.snapConfiguration,
      stretchConfiguration: delegate.stretchConfiguration,
      showOnScreenConfiguration: delegate.showOnScreenConfiguration,
    );
  }

  @override
  void updateRenderObject(BuildContext context,
      _RenderSliverFloatingPinnedPersistentHeaderForWidgets renderObject) {
    renderObject.vsync = delegate.vsync;
    renderObject.snapConfiguration = delegate.snapConfiguration;
    renderObject.stretchConfiguration = delegate.stretchConfiguration;
    renderObject.showOnScreenConfiguration = delegate.showOnScreenConfiguration;
  }
}

class _RenderSliverFloatingPersistentHeaderForWidgets
    extends RenderSliverFloatingPersistentHeader
    with _RenderSliverPersistentHeaderForWidgetsMixin {
  _RenderSliverFloatingPersistentHeaderForWidgets({
    RenderBox? child,
    required TickerProvider? vsync,
    FloatingHeaderSnapConfiguration? snapConfiguration,
    OverScrollHeaderStretchConfiguration? stretchConfiguration,
    PersistentHeaderShowOnScreenConfiguration? showOnScreenConfiguration,
  }) : super(
          child: child,
          vsync: vsync,
          snapConfiguration: snapConfiguration,
          stretchConfiguration: stretchConfiguration,
          showOnScreenConfiguration: showOnScreenConfiguration,
        );
}
